kvbc/src/sparse_merkle/proof_path_processor.cpp

```diff
@@ -0,0 +1,139 @@
+// Concord
+//
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include "sparse_merkle/proof_path_processor.h"
+#include "sparse_merkle/internal_node.h"
+#include "sparse_merkle/keys.h"
+
+namespace concord {
+namespace kvbc {
+namespace sparse_merkle {
+namespace proof_path_processor {
+
+using Sliver = concordUtils::Sliver;
+
+bool verifyProofPath(Sliver key, Sliver value, const std::vector<Hash>& proofPath, const Hash& rootHash) {
+  Hasher hasher{};
+  auto keyHash = hasher.hash(key.data(), key.length());
+  auto valueHash = hasher.hash(value.data(), value.length());
+
+  enum class Direction : std::uint8_t { Left, Right };
+
+  std::vector<Direction> pathOrdering;
+  for (int nibble_index = 0; nibble_index < Hash::MAX_NIBBLES; nibble_index++) {
+    auto current_nibble = keyHash.getNibble(nibble_index).data();
+    auto node_index = BatchedInternalNode::nibbleToIndex(current_nibble);
+    std::vector<Direction> subPath;
+    for (int i = 0; i < BatchedInternalNode::MAX_HEIGHT - 1; i++) {
+      if (BatchedInternalNode::isLeftChild(node_index)) {
+        subPath.push_back(Direction::Left);
+      } else {
+        subPath.push_back(Direction::Right);
+      }
+      auto current_index_opt = BatchedInternalNode::parentIndex(node_index);
+      if (current_index_opt.has_value()) {
+        node_index = current_index_opt.value();
+      } else {
+        break;
+      }
+    }
+    pathOrdering.insert(pathOrdering.end(), subPath.rbegin(), subPath.rend());
+  }
+
+  ConcordAssert(pathOrdering.size() >= proofPath.size());
+
+  size_t index = proofPath.size() - 1;
+  Hash next = valueHash;
+  for (auto peerHash = proofPath.rbegin(); peerHash != proofPath.rend(); peerHash++) {
+    if (pathOrdering[index] == Direction::Right) {
+      next = hasher.parent(proofPath[index], next);
+    } else {
+      next = hasher.parent(next, proofPath[index]);
+    }
+    index--;
+  }
+
+  return next == rootHash;
+}
+
+std::vector<Hash> getProofPath(Sliver key, std::shared_ptr<IDBReader> db, const std::string& custom_prefix) {
+  std::vector<Hash> retVal;
+  Hasher hasher;
+  Hash valueHash{};
+  Hash globalLeaveHash{};
+  auto key_hash = hasher.hash(key.data(), key.length());
```

Comment:
I assume we continue hashing the keys for now, but we will change this later right ?

Reply:
Yes, we have not yet changed the internals of our MPT. It still hashes the keys. This is a task scheduled for the near future.